h/static/scripts/directive/test/annotation-test.js

```diff
@@ -0,0 +1,1377 @@
+      function(done) {
+        sandbox.stub($window, 'confirm').returns(true);
+        fakeAnnotationMapper.deleteAnnotation.returns($q.resolve());
+        controller['delete']().then(function() {
+          assert(fakeAnnotationMapper.deleteAnnotation.calledWith(annotation));
+          done();
+        });
+        $timeout.flush();
+      }
+    );
+
+    it(
+      'doesn\'t call annotationMapper.delete() if the delete is cancelled',
+      function() {
+        sandbox.stub($window, 'confirm').returns(false);
+        assert(fakeAnnotationMapper.deleteAnnotation.notCalled);
+      }
+    );
+
+    it(
+      'flashes a generic error if the server cannot be reached',
+      function(done) {
+        sandbox.stub($window, 'confirm').returns(true);
+        fakeAnnotationMapper.deleteAnnotation.returns($q.reject({
+          status: 0
+        }));
+        controller['delete']().then(function() {
+          assert(fakeFlash.error.calledWith(
+            'Service unreachable.', 'Deleting annotation failed'));
+          done();
+        });
+        $timeout.flush();
+      }
+    );
+
+    it('flashes an error if the delete fails on the server', function(done) {
+      sandbox.stub($window, 'confirm').returns(true);
+      fakeAnnotationMapper.deleteAnnotation.returns($q.reject({
+        status: 500,
+        statusText: 'Server Error',
+        data: {}
+      }));
+      controller['delete']().then(function() {
+        assert(fakeFlash.error.calledWith(
+          '500 Server Error', 'Deleting annotation failed'));
+        done();
+      });
+      $timeout.flush();
+    });
+
+    it('doesn\'t flash an error if the delete succeeds', function(done) {
+      sandbox.stub($window, 'confirm').returns(true);
+      fakeAnnotationMapper.deleteAnnotation.returns($q.resolve());
+      controller['delete']().then(function() {
+        assert(fakeFlash.error.notCalled);
+        done();
+      });
+      $timeout.flush();
+    });
+  });
+
+  describe('saving a new annotation', function() {
+    beforeEach(function() {
+      createDirective();
+      fakeFlash.error = sandbox.stub();
+      controller.action = 'create';
+      annotation.$create = sandbox.stub();
+    });
+
+    it(
+      'emits annotationCreated when saving an annotation succeeds',
+      function(done) {
+        sandbox.spy($rootScope, '$emit');
+        annotation.$create.returns(Promise.resolve());
+        controller.save().then(function() {
+          assert($rootScope.$emit.calledWith('annotationCreated'));
+          done();
+        });
+      }
+    );
+
+    it(
+      'flashes a generic error if the server can\'t be reached',
+      function(done) {
+        annotation.$create.returns(Promise.reject({
+          status: 0
+        }));
+        controller.save().then(function() {
+          assert(fakeFlash.error.calledWith(
+            'Service unreachable.', 'Saving annotation failed'));
+          done();
+        });
+        $timeout.flush();
```

Comment:
Hmm, does `$timeout.flush()` actually work here given that we're using a native Promise, not `$q`. In general, I would be inclined to use native promises rather than `$q` going forwards. Given that we're going to have native promises (or polyfills) anyway since they are part of the web platform, we might as well have only one implementation instead of two.